src/ast/passes/loop_return.cpp

```diff
@@ -0,0 +1,164 @@
+    } else if (loop_return_.value()) {
+      // The value was set within the loop, so return it.
+      ret_stmts.emplace_back(ast_.make_node<Jump>(JumpType::RETURN,
+                                                  get_return_var(stmt.node()),
+                                                  Location(stmt.node().loc)));
+    } else {
+      // The value was not set within the loop, so skip it.
+      ret_stmts.emplace_back(
+          ast_.make_node<Jump>(JumpType::RETURN, Location(stmt.node().loc)));
+    }
+    // Make an if contingent on the return value being set, and break
+    // from this local loop if that is the case.
+    auto *ret_block = ast_.make_node<Block>(std::move(ret_stmts),
+                                            Location(stmt.node().loc));
+    auto *ret_if = ast_.make_node<If>(
+        Expression(get_return_set_var(stmt.node())),
+        ret_block,
+        ast_.make_node<Block>(std::vector<Statement>({}),
+                              Location(stmt.node().loc)),
+        Location(stmt.node().loc));
+    auto *block = ast_.make_node<Block>(
+        std::vector<Statement>({ stmt, ret_if }), Location(stmt.node().loc));
+    stmt = Statement(block); // Replace with the compound block.
+  } else if (auto *jmp = stmt.as<Jump>()) {
+    // This rewrites any return statements in a loop as something that will
+    // set the return value and then break the loop. Combined with the above,
+    // which checks this *outside* the loop and injects a return (possibly
+    // rewritten here if inside an inner loop), we have control flow logic
+    // that will be captured by the automatic context.
+    if (loop_depth_ > 0 && jmp->ident == JumpType::RETURN) {
+      if (!loop_return_.has_value()) {
+        loop_return_ = jmp->return_value.has_value();
+      } else if (loop_return_.value() != jmp->return_value.has_value()) {
+        // We cannot handle this case, but it's a more general error.
+        jmp->addError() << "Return value used in loop is inconsistent";
+        return;
+      }
+      std::vector<Statement> ret_stmts;
+      if (loop_return_.value() && jmp->return_value.has_value()) {
+        ret_stmts.emplace_back(
+            ast_.make_node<AssignVarStatement>(get_return_var(stmt.node()),
+                                               jmp->return_value.value(),
+                                               Location(jmp->loc)));
+      }
+      ret_stmts.emplace_back(ast_.make_node<AssignVarStatement>(
+          get_return_set_var(stmt.node()),
+          ast_.make_node<Integer>(1, Location(jmp->loc)),
+          Location(jmp->loc)));
+      ret_stmts.emplace_back(
+          ast_.make_node<Jump>(JumpType::BREAK, Location(jmp->loc)));
+      auto *block = ast_.make_node<Block>(
+          std::vector<Statement>(std::move(ret_stmts)), Location(jmp->loc));
+      stmt = Statement(block);
+    }
+  }
+}
+
+void LoopReturn::visit(Probe &probe)
+{
```

Comment:
I need to do two things here:

* Clear `loop_return_`
* Apply the same logic below to `SubProg` (factoring this out)

I should also add add tests for these cases, since nothing caught these problems.